tests/test_utils.py

```diff
@@ -11,6 +14,192 @@
+        ],
+    )
+    @mock.patch("pdm.utils.shutil.which", return_value="git")
+    def test_no_git_username_and_email(self, git_patch, no_git_username_and_email_patch):
+        with git_patch:
+            with no_git_username_and_email_patch:
+                assert utils.get_user_email_from_git() == ("", "")
+
+    @mock.patch(
+        "pdm.utils.subprocess.check_output",
+        side_effect=[
+            "username",
+            utils.subprocess.CalledProcessError(-1, ["git", "config", "user.email"], "No email"),
+        ],
+    )
+    @mock.patch("pdm.utils.shutil.which", return_value="git")
+    def test_no_git_email(self, git_patch, no_git_email_patch):
+        with git_patch:
+            with no_git_email_patch:
+                assert utils.get_user_email_from_git() == ("username", "")
+
+    @mock.patch(
+        "pdm.utils.subprocess.check_output",
+        side_effect=[utils.subprocess.CalledProcessError(-1, ["git", "config", "user.name"], "No username"), "email"],
+    )
+    @mock.patch("pdm.utils.shutil.which", return_value="git")
+    def test_no_git_username(self, git_patch, no_git_username_patch):
+        with git_patch:
+            with no_git_username_patch:
+                assert utils.get_user_email_from_git() == ("", "email")
+
+    @mock.patch("pdm.utils.subprocess.check_output", side_effect=["username", "email"])
+    @mock.patch("pdm.utils.shutil.which", return_value="git")
+    def test_git_username_and_email(self, git_patch, git_username_and_email_patch):
+        with git_patch:
+            with git_username_and_email_patch:
+                assert utils.get_user_email_from_git() == ("username", "email")
+
+
+@pytest.mark.parametrize(
+    "given,expected",
+    [
+        ("git@github.com/pdm-project/pdm", "ssh://git@github.com/pdm-project/pdm"),
+        ("ssh://git@github.com/pdm-project/pdm", "ssh://git@github.com/pdm-project/pdm"),
+        ("git+ssh://git@github.com/pdm-project/pdm", "git+ssh://git@github.com/pdm-project/pdm"),
+        ("https://git@github.com/pdm-project/pdm", "https://git@github.com/pdm-project/pdm"),
+        ("file:///my/local/pdm-project/pdm", "file:///my/local/pdm-project/pdm"),
+    ],
+)
+def test_add_ssh_scheme_to_git_uri(given, expected):
+    assert utils.add_ssh_scheme_to_git_uri(given) == expected
+
+
+class TestUrlToPath:
+    def test_non_file_url(self):
+        with pytest.raises(AssertionError):
+            utils.url_to_path("not_a_file_scheme://netloc/path")
+
+    def test_non_windows_non_local_file_url(self):
+        with mock.patch("pdm.utils.sys.platform", "non_windows"):
+            with pytest.raises(ValueError):
+                utils.url_to_path("file://non_local_netloc/file/url")
+
+    def test_non_windows_localhost_local_file_url(self):
+        with mock.patch("pdm.utils.sys.platform", "non_windows"):
+            assert utils.url_to_path("file://localhost/local/file/path") == "/local/file/path"
```

Comment:
Maybe skip it on windows and add another one for Windows only.

Reply:
Np.

Reply:
OK, I've added a single Windows-only test case for `utils.url_to_path` testing the conversion of a localhost local file URL to a Windows path. I hope this works - I don't use Windows and don't have access to a Windows Python environment, so it is tricky for me to test this.